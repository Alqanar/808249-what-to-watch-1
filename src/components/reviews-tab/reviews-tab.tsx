import * as React from "react";
import {connect} from "react-redux";

import {Operation} from "../../reducer/reviews/reviews.js";
import {IReview, IFilm} from "../../types";
import ReviewsTabItem from "../reviews-tab-item/reviews-tab-item";


interface IProps {
  film: IFilm;
  reviews: IReview[];
  loadReviews: (id: string) => Promise<void>;
}

const getReviewsColumn = (reviews: IReview[]): React.ReactElement => {
  const column = reviews.map((review, i): React.ReactElement =>
    <ReviewsTabItem
      review={review}
      key={i}
    />
  );

  return (
    <div className="movie-card__reviews-col">
      {column}
    </div>
  );
};

class ReviewsTab extends React.PureComponent<IProps, null> {
  public constructor(props) {
    super(props);
  }

  public render(): React.ReactElement {
    const {reviews} = this.props;

    const partOneReviews = reviews.splice(0, Math.ceil(reviews.length / 2));
    const partTwoReviews = reviews;

    return (
      <div className="movie-card__reviews movie-card__row">
        {getReviewsColumn(partOneReviews)}
        {getReviewsColumn(partTwoReviews)}
      </div>
    );
  }

  public componentDidMount(): void {
    const {film: {id}, loadReviews} = this.props;

    if (id) {
      loadReviews(id);
    }
  }
}

const mapDispatchToProps = (dispatch): object => ({
  loadReviews: (id): Promise<void> => dispatch(Operation.loadReviews(id))
});

const mapStateToProps = (state, ownProps): void => {
  let reviews = [];
  const reviewsById = state.reviews.comments[ownProps.film.id];
  if (reviewsById) {
    reviews = reviewsById.sort((reviewA, reviewB): number => (
      new Date(reviewB.date).getTime() - new Date(reviewA.date).getTime()
    ));
  }
  return {
    ...ownProps,
    reviews
  };
};


export {ReviewsTab};

export default connect(mapStateToProps, mapDispatchToProps)(ReviewsTab);
