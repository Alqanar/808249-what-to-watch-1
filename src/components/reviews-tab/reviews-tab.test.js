import React from "react";
import renderer from "react-test-renderer";

import {ReviewsTab} from "./reviews-tab.tsx";


const reviewsMockData = [
  {
    id: 1,
    user: {
      id: 4,
      name: `Kate Muir`,
    },
    rating: 8.9,
    comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
    date: `2019-05-08T14:13:56.569Z`
  },
  {
    id: 2,
    user: {
      id: 8,
      name: `Lika Forester`,
    },
    rating: 10,
    comment: `I laughed, cried and was speechless. 3 hours well used. It was a perfect. I will watch it again and again and again. 10 points well deserved!!!`,
    date: `2019-05-13T15:23:56.569Z`
  }
];

const movieMock = {
  backgroundColor: `#ffffff`,
  coverLink: `https://m.media-amazon.com/images/M/MV5BNmE1ZGI2ZDctZDQxNC00MmRkLWJkYTQtNmY3YmI5MzlhOTVkXkEyXkFqcGdeQXVyMjc1NDA2OA@@._V1_SY1000_CR0,0,1498,1000_AL_.jpg`,
  description: `Post World War II, a British colonel and his wife are assigned to live in Hamburg during the post-war reconstruction, but tensions arise with the German who previously owned the house.`,
  director: `James Kent`,
  genre: [`Drama`, `Romance`, `War`],
  id: `0`,
  favorite: false,
  name: `The Aftermath`,
  posterImage: `https://m.media-amazon.com/images/M/MV5BMTk2MDEyNTE5M15BMl5BanBnXkFtZTgwMzY1NDM4NjM@._V1_SY1000_CR0,0,674,1000_AL_.jpg`,
  posterLink: `https://st.kp.yandex.net/images/film_iphone/iphone360_1000125.jpg`,
  trailer: `https://youtu.be/FPv3e2FZOgo`,
  rating: 6.3,
  released: 2019,
  duration: 108,
  scoresCount: 2726,
  starring: [`Keira Knightley`, `Ned Wills`, `Pandora Colin`],
  videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`
};

it(`Reviews tab list correctly renders`, () => {
  const tree = renderer
    .create(
      <ReviewsTab
        film={movieMock}
        reviews={reviewsMockData}
        loadReviews={() => {}}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
