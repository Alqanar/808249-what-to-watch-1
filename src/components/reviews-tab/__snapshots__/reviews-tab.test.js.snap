// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews tab list correctly renders 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Kate Muir
          </cite>
          <time
            className="review__date"
            dateTime="2019-5-8"
          >
            May 8, 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        8.9
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          I laughed, cried and was speechless. 3 hours well used. It was a perfect. I will watch it again and again and again. 10 points well deserved!!!
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Lika Forester
          </cite>
          <time
            className="review__date"
            dateTime="2019-5-13"
          >
            May 13, 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        10
      </div>
    </div>
  </div>
</div>
`;
